src/flat.rs

```diff
@@ -0,0 +1,1139 @@
+//! Image representations for ffi.
+//!
+//! # Usage
+//!
+//! Imagine you want to offer a very simple ffi interface: The caller provides an image buffer and
+//! your program creates a thumbnail from it and dumps that image as `png`. This module is designed
+//! to help you transition from raw memory data to Rust representation.
+//! 
+//! ```no_run
+//! use std::ptr;
+//! use std::slice;
+//! use image::Rgb;
+//! use image::flat::{FlatSamples, MatrixFormat};
+//! use image::imageops::thumbnail;
+//!
+//! #[no_mangle]
+//! pub extern "C" fn store_rgb8_compressed(
+//!     data: *const u8, len: usize,
+//!     format: *const MatrixFormat
+//! )
+//!     -> bool
+//! {
+//!     let samples = unsafe { slice::from_raw_parts(data, len) };
+//!     let format = unsafe { ptr::read(format) };
+//!
+//!     let buffer = FlatSamples {
+//!         samples,
+//!         format,
+//!     };
+//!
+//!     let view = match buffer.as_view::<Rgb<u8>>() {
+//!         Err(_) => return false, // Invalid format.
+//!         Ok(view) => view,
+//!     };
+//!
+//!     thumbnail(&view, 64, 64)
+//!         .save("output.png")
+//!         .map(|_| true)
+//!         .unwrap_or_else(|_| false)
+//! }
+//! ```
+//! 
+use std::cmp;
+use std::ops::Deref;
+use std::marker::PhantomData;
+
+use num_traits::Zero;
+
+use buffer::{ImageBuffer, Pixel};
+use color::ColorType;
+use image::{GenericImage, GenericImageView, ImageError};
+
+/// A flat buffer over a (multi channel) image.
+///
+/// Note that the strides need not conform to the assumption that constructed
+/// indices actually refer inside the underlying buffer but return values
+/// of library functions will always guarantee this. To manually make this
+/// check use `check_index_validities` and maybe put that inside an assert.
+#[derive(Clone, Debug)]
+pub struct FlatSamples<Buffer> {
+    /// Underlying linear container holding sample values.
+    pub samples: Buffer,
+
+    /// A `repr(C)` description of the buffer format.
+    pub format: MatrixFormat,
+}
```

Comment:
How would you feel about also including and `Option<ColorType>` field for when the image format is known and is one of the supported ones?

Reply:
I'm torn on this. This feels like it deserved an additional consistency check when calling `as_view` but you don't generally gain help in the form of extra type checking. A `into_dynamic_image` would sure make this much more useful.

Reply:
I'll add a `color_hint: Option<ColorType>` with an explicit warning that it is not checked in any of the conversions. A later dynamic converter could then be `into_hinted_image`.